crates/driver/src/domain/competition/solution/slippage.rs

```diff
@@ -0,0 +1,269 @@
+                cmp::min(relative, self.max.unwrap_or(U256::max_value())),
+            );
+
+            tracing::debug!(
+                input_token = ?interaction.input.token,
+                ?relative,
+                ?slippage,
+                "using liquidity output token for capped surplus",
+            );
+
+            // Convert back to input token amount
+            price
+                .from_eth(eth::Ether(slippage))
+                .checked_mul(&interaction.input.amount)
+                .ok_or(super::error::Math::Overflow)?
+                .checked_div(&interaction.output.amount)
+                .ok_or(super::error::Math::DivisionByZero)?
+        } else {
+            tracing::warn!(
+                input_token = ?interaction.input.token,
+                output_token = ?interaction.output.token,
+                "unable to compute capped slippage; falling back to relative slippage",
+            );
+            let relative = interaction.input.amount.0.to_big_rational() * &self.relative;
+            big_rational_to_u256(&relative)
+                .map_err(|_| super::error::Math::DivisionByZero)?
+                .into()
+        };
+
+        tracing::debug!(?interaction, ?slippage, "applying slippage to liquidity",);
+        Ok((
+            MaxInput(eth::Asset {
+                amount: interaction.input.amount + slippage,
+                ..interaction.input
+            }),
+            ExactOutput(interaction.output),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::domain::eth::Asset,
+        num::rational::Ratio,
+        testlib::tokens::{GNO, USDC},
+    };
+
+    #[test]
+    fn test_input_price() {
+        let interaction = Interaction {
+            input: Asset {
+                token: GNO.into(),
+                // 1GNO
+                amount: eth::U256::exp10(18).into(),
+            },
+            output: Asset {
+                token: USDC.into(),
+                // 200 USDC
+                amount: (U256::from(200) * eth::U256::exp10(6)).into(),
+            },
+        };
+        let prices = maplit::hashmap! {
+            // 0.1 ETH
+            GNO.into() => eth::U256::exp10(17).into(),
+        };
+
+        // no cap
+        let slippage = Parameters {
+            relative: Ratio::from_float(0.1).unwrap(),
+            max: None,
+            min: None,
+            prices,
+        };
+        let (max_input, exact_output) = slippage.apply_to(&interaction).unwrap();
+        assert_eq!(
+            max_input.0.amount,
+            (eth::U256::exp10(18) + eth::U256::exp10(17)).into()
+        );
+        assert_eq!(exact_output.0.amount, interaction.output.amount);
+
+        // min cap
+        let min_cap = Parameters {
+            // 0.1 ETH (1 GNO)
+            min: Some(eth::U256::exp10(17)),
+            ..slippage.clone()
+        };
+        let (max_input, _) = min_cap.apply_to(&interaction).unwrap();
```

Comment:
we might as well also assert that the output amount is unchanged we the min and max cap.